mgmtfn/k8splugin/driver.go

```diff
@@ -0,0 +1,370 @@
+			return link, err
+		}
+		// try once more as sometimes (somehow) link creation is taking
+		// sometime, causing link not found error
+		time.Sleep(1 * time.Second)
+		link, err = netlink.LinkByName(ifname)
+		if err != nil {
+			log.Errorf("unable to find link %q. Error %q", ifname, err)
+		}
+		return link, err
+	}
+	return link, err
+}
+
+// nsToPID is a utility that extracts the PID from the netns
+func nsToPID(ns string) (int, error) {
+	// Make sure ns is well formed
+	ok := strings.HasPrefix(ns, "/proc/")
+	if !ok {
+		return -1, fmt.Errorf("Invalid nw name space: %v", ns)
+	}
+
+	elements := strings.Split(ns, "/")
+	return strconv.Atoi(elements[2])
+}
+
+// setIfAttrs sets the required attributes for the container interface
+func setIfAttrs(ifname, netns, cidr, newname string) error {
+
+	// convert netns to pid that netlink needs
+	pid, err := nsToPID(netns)
+	if err != nil {
+		return err
+	}
+
+	nsenterPath, err := osexec.LookPath("nsenter")
+	if err != nil {
+		return err
+	}
+	ipPath, err := osexec.LookPath("ip")
+	if err != nil {
+		return err
+	}
+
+	// find the link
+	link, err := getLink(ifname)
+	if err != nil {
+		log.Errorf("unable to find link %q. Error %q", ifname, err)
+		return err
+	}
+
+	// move to the desired netns
+	err = netlink.LinkSetNsPid(link, pid)
+	if err != nil {
+		log.Errorf("unable to move interface %s to pid %d. Error: %s",
+			ifname, pid, err)
+		return err
+	}
+
+	// rename to the desired ifname
+	nsPid := fmt.Sprintf("%d", pid)
+	rename, err := osexec.Command(nsenterPath, "-t", nsPid, "-n", "-F", "--", ipPath, "link",
+		"set", "dev", ifname, "name", newname).CombinedOutput()
+	if err != nil {
+		log.Errorf("unable to rename interface %s to %s. Error: %s",
+			ifname, newname, err)
+		return nil
+	}
+	log.Infof("Output from rename: %v", rename)
+
+	// set the ip address
+	assignIP, err := osexec.Command(nsenterPath, "-t", nsPid, "-n", "-F", "--", ipPath,
+		"address", "add", cidr, "dev", newname).CombinedOutput()
+
+	if err != nil {
+		log.Errorf("unable to assign ip %s to %s. Error: %s",
+			cidr, newname, err)
+		return nil
+	}
+	log.Infof("Output from ip assign: %v", assignIP)
+
+	// Finally, mark the link up
+	bringUp, err := osexec.Command(nsenterPath, "-t", nsPid, "-n", "-F", "--", ipPath,
+		"link", "set", "dev", newname, "up").CombinedOutput()
+
+	if err != nil {
+		log.Errorf("unable to assign ip %s to %s. Error: %s",
+			cidr, newname, err)
+		return nil
+	}
+	log.Infof("Output from ip assign: %v", bringUp)
+	return nil
+
+}
+
+// getEPSpec gets the EP spec using the pod attributes
+func getEPSpec(pInfo *cniapi.CNIPodAttr) (*epSpec, error) {
+	resp := epSpec{}
+
+	// Get labels from the kube api server
+	epg, err := kubeAPIClient.GetPodLabel(pInfo.K8sNameSpace, pInfo.Name, "net-group")
+	if err != nil {
+		log.Errorf("Error getting epg. Err: %v", err)
+		return &resp, err
+	}
+
+	netw, _ := kubeAPIClient.GetPodLabel(pInfo.K8sNameSpace, pInfo.Name, "network")
```

Comment:
Actually, if the first check passes the subsequent ones are guaranteed to pass. I will add a comment to that effect.